yarpl/include/yarpl/flowable/FlowableOperator.h

```diff
@@ -466,6 +470,379 @@ class FromPublisherOperator : public Flowable<T> {
   OnSubscribe function_;
 };
 
+template <typename T, typename R>
+class FlatMapOperator : public FlowableOperator<T, R, FlatMapOperator<T, R>> {
+  using ThisOperatorT = FlatMapOperator<T, R>;
+  using Super = FlowableOperator<T, R, ThisOperatorT>;
+
+ public:
+  FlatMapOperator(
+      Reference<Flowable<T>> upstream,
+      folly::Function<Reference<Flowable<R>>(T)> func)
+      : Super(std::move(upstream)), function_(std::move(func)) {}
+
+  void subscribe(Reference<Subscriber<R>> subscriber) override {
+    Super::upstream_->subscribe(make_ref<FMSubscription>(
+        this->ref_from_this(this), std::move(subscriber)));
+  }
+
+ private:
+  using SuperSubscription = typename Super::Subscription;
+  class FMSubscription : public SuperSubscription {
+    struct MappedStreamSubscriber;
+
+   public:
+    FMSubscription(
+        Reference<ThisOperatorT> flowable,
+        Reference<Subscriber<R>> subscriber)
+        : SuperSubscription(std::move(flowable), std::move(subscriber)) {}
+
+    void onSubscribeImpl() final {
+      liveSubscribers_++;
+      SuperSubscription::onSubscribeImpl();
+      // TODO: make max parallelism configurable a-la RxJava 2.x's
+      // FlowableFlatMapOperator
+      SuperSubscription::request(std::numeric_limits<int64_t>::max());
+    }
+
+    void onNextImpl(T value) final {
+      auto&& flatMapOp = this->getFlowableOperator();
```

Comment:
should we check if the Subscription already got an exception or closed?

Reply:
That'll happen on each iteration of the work loop; we want to do checks like that in there so we know all work is serialized

Reply:
The next iteration of drainImpl() will take care of the cancel() or onError cases; because of the async nature of flatMap, it's unknown if a mapping subscriber or the upstream will throw an error first, so we serialize in drainImpl()